src/internal.rs

```diff
@@ -272,6 +272,43 @@ pub(crate) struct TaskThreadData {
     pub delayed_fg: RwLock<TaskThreadDataDelayedFg>,
 }
 
+impl TaskThreadData {
+    /// When we finish a frame we advance `first`, and we must also adjust `cur`,
+    /// because `cur` is an offset of `first`.
+    /// We may also need to adjust `reset_task_cur`.
+    ///
+    /// The task thread lock must be held and there must be more than 1 frame context.
+    pub(crate) fn advance_first(&self, n_fc: usize) {
+        // This read-modify-write doesn't need to be an atomic CAS:
+        // modifications are protected by the lock that we're holding.
```

Comment:
If we already have a lock for this, why is `SeqCst` needed?

Reply:
Especially because the `.load` and `.fetch_add(1` aren't atomic together.

Reply:
Don't need to fix this right now, but I'm wondering.

Reply:
It's a good question. At the moment it's a direct move of the existing code. I did have a commit which I didn't include which made it Relaxed, but I didn't include it because it starts to get quite complex to tease out what is and isn't an access which might require a more rigorous ordering.

I'll have another look at including that commit in the series; the other thing I wanted to compare was putting `cur`, `reset_task_cur` and `first` in a lock all of their own so that there was just one single acquire/release pair of atomic operations, and it would simplify a lot of the code.